### PR TITLE
Allow configuring a custom DNS resolver for the JDBC driver

### PR DESCRIPTION
**Pull request by aalbu**
## Description

Provide a way to configure a custom DNS resolver for the JDBC driver.

Adds 2 new new JDBC URL parameters:

- `dnsResolver` - the name of the class implementing interface `io.trino.jdbc.DnsResolver`
- `dnsResolverContext` - a string argument to be passed to the constructor of the above class (if this parameter is not provided, `null` will be used)

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(X) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown

```

### CODE REVIEW

**Review by hashhar — CHANGES_REQUESTED** [review]
why? what is the goal which this change helps accomplish?

DNS must come from the host system. I don't know of any other JDBC driver which allows this. This can also lead to possibility of different DNS in JDBC and within the JVM which hosts the JDBC driver.

**Comment by thepetlyon** [comment]
Hey Ashhar! I've been working with Alex on the motivation behind this change. Essentially, we are working with a use case where we are hosting a Trino instance inside of a VPC that may need to work with DNS resolution inside of a private subnet. In our case, we need the ability to inject a custom DNS resolver so that when Trino JDBC Driver [fetches the paginated query results](https://trino.io/docs/current/develop/client-protocol.html), it will be able to perform a proper DNS query on the returned nextUri if this nextUri needs to be resolved to IP addresses from within the VPC.

**Review by thepetlyon — APPROVED** [review]
Looks great Alex! Thank you so much for tackling this issue for us!

File: client/trino-jdbc/src/main/java/io/trino/jdbc/DnsResolver.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ */
+package io.trino.jdbc;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+
+public interface DnsResolver
+{
+    List<InetAddress> lookup(String hostname) throws UnknownHostException;
```
Comment (electrum) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/DnsResolver.java:22]: Formatting

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java [hunk 2]
```diff
@@ -348,6 +350,9 @@ public void setupClient(OkHttpClient.Builder builder)
                 builder.authenticator(authenticator);
                 builder.addInterceptor(authenticator);
             }
+
+            Optional<String> resolverContext = DNS_RESOLVER_CONTEXT.getValue(properties);
+            DNS_RESOLVER.getValue(properties).ifPresent(resolverClass -> builder.dns(new OkHttpDnsAdapter(instantiateDnsResolver(resolverClass, resolverContext))));
```
Comment (electrum) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:355]: Do we need the class `OkHttpDnsAdapter`? I think we could do
```java
builder.dns(instantiateDnsResolver(resolverClass, resolverContext)::lookup)

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java [hunk 3]
```diff
@@ -357,6 +362,18 @@ public void setupClient(OkHttpClient.Builder builder)
         }
     }
 
+    private static DnsResolver instantiateDnsResolver(Class<? extends DnsResolver> resolverClass, Optional<String> context)
+    {
+        try {
+            return context.isPresent() ?
```
Comment (electrum) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:368]: Let's simplify and require the single argument constructor. Using a different method signature based on whether or not the user provided a context argument seems strange. We can pass null if the context is not present.

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java [hunk 4]
```diff
@@ -357,6 +362,18 @@ public void setupClient(OkHttpClient.Builder builder)
...
+                resolverClass.getConstructor(String.class).newInstance(context.get()) :
+                resolverClass.getDeclaredConstructor().newInstance();
+        }
+        catch (Exception e) {
```
Comment (electrum) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:372]: Catch `ReflectiveOperationException` here

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java [hunk 5]
```diff
@@ -357,6 +362,18 @@ public void setupClient(OkHttpClient.Builder builder)
...
+            throw new RuntimeException("Unable to instantiate custom DNS resolver " + resolverClass.getName(), e);
```
Comment (electrum) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:373]: Nit: add colon after error message

Comment (electrum) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:373]: Throw `ClientException` here so that the original message is returned to the user

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java [hunk 6]
```diff
@@ -1084,6 +1091,18 @@ public void testUpdatePartialCancel()
         }
     }
 
+    @Test(expectedExceptions = SQLFeatureNotSupportedException.class, expectedExceptionsMessageRegExp = "Multiple open results not supported")
```
Comment (electrum) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:1094]: Don't rearrange existing tests, as that is unrelated to the DNS resolver change

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java [hunk 7]
```diff
@@ -604,14 +607,18 @@ public void testGetResultSet()
...
     }
 
-    @Test(expectedExceptions = SQLFeatureNotSupportedException.class, expectedExceptionsMessageRegExp = "Multiple open results not supported")
-    public void testGetMoreResultsException()
+    @Test
+    public void testCustomDnsResolver()
             throws Exception
     {
-        try (Connection connection = createConnection()) {
+        Properties props = new Properties();
```
Comment (electrum) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:614]: Nit: name this `properties`

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java [hunk 8]
```diff
@@ -604,14 +607,18 @@ public void testGetResultSet()
...
+        props.setProperty("dnsResolver", TestDnsResolver.class.getName());
+        props.setProperty("dnsResolverContext", server.getAddress().getHost());
+        props.setProperty("user", "test");
+
+        try (Connection connection = DriverManager.getConnection("jdbc:trino://mycustomaddress:" + server.getAddress().getPort(), props)) {
```
Comment (electrum) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:619]: Let's extract a variable for the URL

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java [hunk 9]
```diff
@@ -1144,4 +1163,26 @@ private static Properties toProperties(Map<String, String> map)
         map.forEach(properties::setProperty);
         return properties;
     }
+
+    public static class TestDnsResolver
```
Comment (electrum) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:1167]: Name this `TestingDnsResolver` since the `Test` prefix is reserved for classes that have tests, whereas this is an implementation for something "for testing"

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java [hunk 10]
```diff
@@ -1144,4 +1163,26 @@ private static Properties toProperties(Map<String, String> map)
...
+            implements DnsResolver
+    {
+        private final InetAddress address;
+
+        public TestDnsResolver(String context)
+                throws UnknownHostException
+        {
+            this.address = InetAddress.getByName(requireNonNull(context, "context is null"));
```
Comment (electrum) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:1175]: It would be better to do the resolution in the `lookup()` method, since that would be a more typical implementation

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java [hunk 11]
```diff
@@ -1144,4 +1163,26 @@ private static Properties toProperties(Map<String, String> map)
...
+        }
+
+        @Override
+        public List<InetAddress> lookup(String hostname)
+                throws UnknownHostException
+        {
+            if ("mycustomaddress".equals(hostname)) {
+                return ImmutableList.of(address);
+            }
+            throw new UnknownHostException(hostname + " could not be resolved");
```
Comment (electrum) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:1185]:
```java
throw new UnknownHostException("Cannot resolve host: " + hostname);

File: client/trino-jdbc/src/main/java/io/trino/jdbc/OkHttpDnsAdapter.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+import okhttp3.Dns;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class OkHttpDnsAdapter
```
Comment (electrum) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/OkHttpDnsAdapter.java:24]: We shouldn't add this as a public class in the driver, but actually, it's probably not needed at all. See comment at usage site.

**Review by electrum — COMMENTED** [review]
Generally looks good

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java (hunk 2 above)
Comment (aalbu) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:355]: Yes, thanks.

File: client/trino-jdbc/src/main/java/io/trino/jdbc/OkHttpDnsAdapter.java (hunk 12 above)
Comment (aalbu) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/OkHttpDnsAdapter.java:24]: Good point.

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java (hunk 6 above)
Comment (aalbu) [reply client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:1094]: I just added a new test, but this is how GH rendered the diff.  I've moved it as the last test in the class.

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java [hunk 13]
```diff
@@ -357,6 +362,16 @@ public void setupClient(OkHttpClient.Builder builder)
         }
     }
 
+    private static DnsResolver instantiateDnsResolver(Class<? extends DnsResolver> resolverClass, Optional<String> context)
+    {
+        try {
+            return resolverClass.getConstructor(String.class).newInstance(context.orElse(null));
+        }
+        catch (ReflectiveOperationException e) {
+            throw new ClientException("Unable to instantiate custom DNS resolver " + resolverClass.getName() + ".", e);
```
Comment (electrum) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:371]:
```java
throw new ClientException("Unable to instantiate custom DNS resolver: " + resolverClass.getName(), e);

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java [hunk 14]
```diff
@@ -1144,4 +1164,25 @@ private static Properties toProperties(Map<String, String> map)
...
+        }
+
+        @Override
+        public List<InetAddress> lookup(String hostname)
+                throws UnknownHostException
+        {
+            if ("mycustomaddress".equals(hostname)) {
+                return ImmutableList.of(InetAddress.getByName(context));
+            }
+            throw new UnknownHostException("Cannot resolve host: " + hostname + ".");
```
Comment (electrum) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:1185]: Nit: remove period from end

File: client/trino-jdbc/src/main/java/io/trino/jdbc/ConnectionProperties.java [hunk 15]
```diff
@@ -613,6 +617,34 @@ public Source()
...
+            super("dnsResolver", NOT_REQUIRED, ALLOWED, Resolver::findByName);
+        }
+
+        public static Class<? extends DnsResolver> findByName(String name)
+        {
+            try {
+                return Class.forName(name).asSubclass(DnsResolver.class);
+            }
+            catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
```
Comment (electrum) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/ConnectionProperties.java:634]: How about
```java
throw new RuntimeException("DNS resolver class not found: " + name, e);

**Review by electrum — APPROVED** [review]
